plugins/aws-catalog-backend/src/BaseEntityProvider.ts

```diff
@@ -0,0 +1,133 @@
+  protected id?: string;
+
+  private readonly transformer: InstanceTransformer<T>;
+  private readonly filter: InstanceFilter<T>;
+
+  constructor(options: {
+    id?: string;
+    logger: Logger;
+    taskRunner: TaskRunner;
+    transformer: InstanceTransformer<T>;
+    filter: InstanceFilter<T> | undefined;
+  }) {
+    this.logger = options.logger.child({
+      target: this.getProviderName(),
+    });
+    this.scheduleFn = this.createScheduleFn(options.taskRunner);
+    this.id = options.id;
+    this.transformer = options.transformer;
+    this.filter = options.filter || (() => true);
+  }
+
+  getProviderName(): string {
+    const suffix = this.id ? `:${this.id}` : '';
+    return `${this.constructor.name}${suffix}`;
+  }
+
+  async connect(connection: EntityProviderConnection): Promise<void> {
+    this.logger.debug('Connected to the catalog');
+    this.connection = connection;
+    await this.scheduleFn();
+  }
+
+  private createScheduleFn(taskRunner: TaskRunner): () => Promise<void> {
+    return async () => {
+      const taskId = `${this.getProviderName()}:refresh`;
+      return taskRunner.run({
+        id: taskId,
+        fn: async () => {
+          const logger = this.logger.child({
+            class: this.constructor.name,
+            taskId,
+            taskInstanceId: uuid.v4(),
+          });
+
+          try {
+            await this.refresh(logger);
+          } catch (error) {
+            logger.error(
+              `${this.getProviderName()} refresh failed, ${error}`,
+              error,
+            );
+          }
+        },
+      });
+    };
+  }
+
+  private async refresh(logger: Logger): Promise<void> {
+    logger.debug(`Refreshing ${this.getProviderName()}`);
+
+    if (!this.connection) {
+      throw new Error(
+        `Connection not initialized for ${this.getProviderName()}`,
+      );
+    }
+
+    const instances = await this.getInstances();
+
+    logger.info(`Retrieved ${instances.length} instances from AWS.`);
+
+    const entities: ResourceEntity[] = await Promise.all(
+      instances.filter(this.filter).map(this.createResource),
+    );
+
+    logger.info(`Will apply ${entities.length} instances to the catalog.`);
+
+    await this.connection.applyMutation({
+      type: 'full',
+      entities: entities.map(entity => ({
+        locationKey: this.getProviderName(),
+        entity: entity,
+      })),
+    });
+  }
+
+  protected abstract getInstances(): Promise<T[]>;
+
+  protected abstract getArn(resource: T): string;
+
+  private createResource = async (instance: T): Promise<ResourceEntity> => {
+    const resourceEntity = await this.transformer(instance);
+    const instanceArn = this.getArn(instance);
+
+    return merge(
+      {
+        metadata: {
+          annotations: {
+            [arn.ANNOTATION]: instanceArn,
+            [ANNOTATION_LOCATION]: instanceArn,
+            [ANNOTATION_ORIGIN_LOCATION]: instanceArn,
+          },
+        },
+      },
```

Comment:
not for this change but i'm trying to better understand how refreshing on a per entity level works. in the out of the box `AboutCard` there's a guard against refreshing anything with a location that's not `url:` or `file:` (not for any reason we have to worry about with `arn:` locations)

https://github.com/backstage/backstage/blob/ebe7afad9d19f279469168ca0d4feceb92c1ad36/plugins/catalog/src/components/AboutCard/AboutCard.tsx#L201-L215

ui aside, i'm a little confused how refreshing an individual entity works. (maybe has to do with these `readLocation` on a processor, which is separate from the provider: https://backstage.io/docs/features/software-catalog/external-integrations#creating-a-catalog-data-reader-processor.)

anyway, at some point we'll want to be able to refresh one of our dbs via the ui i'm sure, so wanted to flag that here